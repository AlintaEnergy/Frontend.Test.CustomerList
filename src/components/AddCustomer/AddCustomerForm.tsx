import { Field, Formik, FormikHelpers } from 'formik';
import * as React from 'react';
import { ICustomer, Customer } from '../../types/types';
import * as Yup from 'yup';
import {
    StyledForm,
    StyledInput,
    StyledLabel,
    StyledAddButton,
    StyledErrorMessage,
} from './StyledAddCustomerForm';
import { NavLink } from 'react-router-dom';

const AddCustomerSchema = Yup.object().shape({
    firstName: Yup.string().required('Required'),
    lastName: Yup.string().required('Required'),
    phoneNumber: Yup.string()
        .matches(/^(?:\+61|0)[1-9](?:[0-9]?[0-9])?[0-9]{6}$/, {
            message: 'Invalid phone number',
            excludeEmptyString: true,
        })
        .required('Required'),
    birthday: Yup.date().max(new Date(), 'Invalid date'),
});

type Props = {
    saveCustomer: (customer: ICustomer | any) => void;
};

export const AddCustomerForm: React.FC<Props> = ({ saveCustomer }) => {
    const [message, setMessage] = React.useState('');

    return (
        <Formik
            initialValues={{
                firstName: '',
                lastName: '',
                phoneNumber: '',
            }}
            validationSchema={AddCustomerSchema}
            onSubmit={(
                values: Customer,
                { setSubmitting }: FormikHelpers<Customer>
            ) => {
                saveCustomer(values);
                setSubmitting(false);
                setMessage('User added');
            }}
        >
            {({ errors, touched }) => (
                <StyledForm>
                    <StyledLabel htmlFor='firstName'>First Name</StyledLabel>
                    <Field
                        as={StyledInput}
                        id='firstName'
                        name='firstName'
                        placeholder='John'
                    />

                    {errors.firstName && touched.firstName ? (
                        <StyledErrorMessage>
                            {errors.firstName}
                        </StyledErrorMessage>
                    ) : null}

                    <StyledLabel htmlFor='lastName'>Last Name</StyledLabel>
                    <Field
                        as={StyledInput}
                        id='lastName'
                        name='lastName'
                        placeholder='Doe'
                    />

                    {errors.lastName && touched.lastName ? (
                        <StyledErrorMessage>
                            {errors.lastName}
                        </StyledErrorMessage>
                    ) : null}

                    <StyledLabel htmlFor='phoneNumber'>
                        Phone Number
                    </StyledLabel>
                    <Field
                        as={StyledInput}
                        id='phoneNumber'
                        name='phoneNumber'
                        placeholder='0412345678'
                        type='tel'
                    />

                    {errors.phoneNumber && touched.phoneNumber ? (
                        <StyledErrorMessage>
                            {errors.phoneNumber}
                        </StyledErrorMessage>
                    ) : null}
                    <StyledLabel htmlFor='birthday'>Birthday</StyledLabel>
                    <Field
                        as={StyledInput}
                        id='birthday'
                        name='birthday'
                        placeholder='01/01/2000'
                        type='date'
                    />
                    {errors.birthday && touched.birthday ? (
                        <StyledErrorMessage>
                            {errors.birthday}
                        </StyledErrorMessage>
                    ) : null}

                    <span style={{ color: 'green' }}>{message}</span>
                    <div
                        style={{
                            display: 'flex',
                            justifyContent: 'space-evenly',
                        }}
                    >
                        <NavLink to='/'>
                            <StyledAddButton type='button'>
                                Back
                            </StyledAddButton>
                        </NavLink>

                        <StyledAddButton type='submit'>
                            Add Customer
                        </StyledAddButton>
                    </div>
                </StyledForm>
            )}
        </Formik>
    );
};
