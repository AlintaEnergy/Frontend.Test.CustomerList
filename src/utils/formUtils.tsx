// eslint-disable-next-line no-useless-escape
export const nameRegex = /^[A-Za-z\s\'-]*$/;

// Accepting 04xx xxx xxx, (0x) xxxx xxxx, +61 xxx xxx xxx and without space
// eslint-disable-next-line no-useless-escape
export const phoneRegex = /^(?:\+?(61))? ?(?:\((?=.*\)))?(0?[2-57-8])\)? ?(\d\d(?:[- ](?=\d{3})|(?!\d\d[- ]?\d[- ]))\d\d[- ]?\d[- ]?\d{3})$/;

export const toHeaderCase = (input: string) => {
  const replaceInput = input.replace(/([A-Z])/g, " $1");
  const changedInput =
    replaceInput.charAt(0).toUpperCase() + replaceInput.slice(1);
  return changedInput;
};
