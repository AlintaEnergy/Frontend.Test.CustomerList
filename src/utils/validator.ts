export const validateEmptyField = (value:string, errorMessage: string) => {
    return !value ? errorMessage : undefined 
  }

export const validatePattern = (value: string, pattern: RegExp, errorMessage: string) => {
    return !pattern.test(value) ? errorMessage : undefined
  }