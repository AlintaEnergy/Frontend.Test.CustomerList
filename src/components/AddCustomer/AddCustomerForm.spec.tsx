import * as React from 'react';
import { render } from '../../utils/testUtils';
import { AddCustomerForm } from './AddCustomerForm';
import { BrowserRouter } from 'react-router-dom';

describe('<AddCustomerForm />', () => {
    it('should render a <AddCustomerForm />', () => {
        const wrapper = render(
            <BrowserRouter>
                <AddCustomerForm saveCustomer={() => {}} />
            </BrowserRouter>
        );
        expect(wrapper.container).toMatchSnapshot();
    });

    //@TODO Add tests for entering data and clicking submit
});
