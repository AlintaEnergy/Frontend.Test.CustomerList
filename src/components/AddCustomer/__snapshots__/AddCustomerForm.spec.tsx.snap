// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddCustomerForm /> should render a <AddCustomerForm /> 1`] = `
<div>
  <form
    action="#"
    class="sc-aXZVg NgUjJ"
  >
    <label
      class="sc-gEvEer cJUhzO"
      for="firstName"
    >
      First Name
    </label>
    <input
      class="sc-eqUAAy BXhUo"
      id="firstName"
      name="firstName"
      placeholder="John"
      value=""
    />
    <label
      class="sc-gEvEer cJUhzO"
      for="lastName"
    >
      Last Name
    </label>
    <input
      class="sc-eqUAAy BXhUo"
      id="lastName"
      name="lastName"
      placeholder="Doe"
      value=""
    />
    <label
      class="sc-gEvEer cJUhzO"
      for="phoneNumber"
    >
      Phone Number
    </label>
    <input
      class="sc-eqUAAy BXhUo"
      id="phoneNumber"
      name="phoneNumber"
      placeholder="0412345678"
      type="tel"
      value=""
    />
    <label
      class="sc-gEvEer cJUhzO"
      for="birthday"
    >
      Birthday
    </label>
    <input
      class="sc-eqUAAy BXhUo"
      id="birthday"
      name="birthday"
      placeholder="01/01/2000"
      type="date"
      value=""
    />
    <span
      style="color: green;"
    />
    <div
      style="display: flex; justify-content: space-evenly;"
    >
      <a
        aria-current="page"
        class="active"
        href="/"
      >
        <button
          class="sc-fqkvVR igBLMC"
          type="button"
        >
          Back
        </button>
      </a>
      <button
        class="sc-fqkvVR igBLMC"
        type="submit"
      >
        Add Customer
      </button>
    </div>
  </form>
</div>
`;
